### PR TITLE
Use repartition in window functions to speed up

### LINKED ISSUE #435: create a test with more than one partition for window functions
I think it would be valuable to create a test with more than one partition (so that the data may not arrive to the `WindowFunction` as a single `RecordBatch`)

__Originally posted by @alamb in https://github.com/apache/arrow-datafusion/pull/403#discussion_r640840946__

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

this pull request is based on #571 so review that first

Closes #435 as this will make it stale

 # Rationale for this change

benchmark:

```

window empty over, aggregate functions
                        time:   [7.6349 ms 7.6997 ms 7.7675 ms]
                        change: [-78.341% -78.072% -77.798%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

window empty over, built-in functions
                        time:   [6.5445 ms 6.5839 ms 6.6256 ms]
                        change: [-78.729% -78.298% -77.872%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

Benchmarking window order by, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 18.3s, or reduce sample count to 20.
window order by, aggregate functions
                        time:   [177.92 ms 179.66 ms 181.50 ms]
                        change: [-10.040% -8.7722% -7.3530%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking window order by, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 16.8s, or reduce sample count to 20.
window order by, built-in functions
                        time:   [165.02 ms 166.82 ms 168.68 ms]
                        change: [-12.674% -11.085% -9.5879%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking window partition by, u64_wide, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 27.7s, or reduce sample count to 10.
window partition by, u64_wide, aggregate functions
                        time:   [268.22 ms 271.51 ms 274.78 ms]
                        change: [-8.0164% -6.5188% -5.0558%] (p = 0.00 < 0.05)
                        Performance has improved.

Benchmarking window partition by, u64_narrow, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.2s, or reduce sample count to 40.
window partition by, u64_narrow, aggregate functions
                        time:   [116.73 ms 118.42 ms 120.14 ms]
                        change: [-22.910% -21.502% -20.118%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking window partition by, u64_wide, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 24.2s, or reduce sample count to 20.
window partition by, u64_wide, built-in functions
                        time:   [229.51 ms 231.90 ms 234.33 ms]
                        change: [-7.9804% -6.7551% -5.5555%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking window partition by, u64_narrow, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 11.0s, or reduce sample count to 40.
window partition by, u64_narrow, built-in functions
                        time:   [105.30 ms 106.75 ms 108.24 ms]
                        change: [-23.288% -21.785% -20.243%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking window partition and order by, u64_wide, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 47.0s, or reduce sample count to 10.
window partition and order by, u64_wide, aggregate functions
                        time:   [441.09 ms 444.76 ms 448.45 ms]
                        change: [-12.907% -11.959% -10.947%] (p = 0.00 < 0.05)
                        Performance has improved.

Benchmarking window partition and order by, u64_narrow, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 57.2s, or reduce sample count to 10.
window partition and order by, u64_narrow, aggregate functions
                        time:   [563.99 ms 568.03 ms 572.04 ms]
                        change: [-15.042% -13.897% -12.764%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking window partition and order by, u64_wide, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 41.4s, or reduce sample count to 10.
window partition and order by, u64_wide, built-in functions
                        time:   [400.88 ms 403.92 ms 406.91 ms]
                        change: [-3.1220% -1.9172% -0.6254%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild

Benchmarking window partition and order by, u64_narrow, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 34.2s, or reduce sample count to 10.
window partition and order by, u64_narrow, built-in functions
                        time:   [334.65 ms 337.16 ms 339.72 ms]
                        change: [-35.596% -33.771% -31.982%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/benches/window_query_sql.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    let mut ctx = ExecutionContext::new();
+    let provider = create_table_provider(partitions_len, array_len, batch_size)?;
+    ctx.register_table("t", provider)?;
+    Ok(Arc::new(Mutex::new(ctx)))
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    let partitions_len = 8;
+    let array_len = 32768 * 2; // 2^16
+    let batch_size = 2048; // 2^11
```
Comment (Dandandan) [inline datafusion/benches/window_query_sql.rs:51]: I think this generates 32 batches divided over 8 partitions total, right?
I think we should (also) benchmark with a slightly larger set of e.g. 1M records in total.

Also batch size could be set to use the default (8192 now) to have a setting that corresponds more to current configuration.

**Comment by Jimexist** [comment]
benchmark run

```
Benchmarking window empty over, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.9s, enable flat sampling, or reduce sample count to 60.
window empty over, aggregate functions
                        time:   [1.1699 ms 1.1755 ms 1.1811 ms]
                        change: [-37.629% -36.780% -36.002%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

Benchmarking window empty over, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.3s, enable flat sampling, or reduce sample count to 60.
window empty over, built-in functions
                        time:   [1.0180 ms 1.0236 ms 1.0299 ms]
                        change: [-26.237% -25.544% -24.778%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

window order by, aggregate functions
                        time:   [8.0163 ms 8.0741 ms 8.1409 ms]
                        change: [-3.3019% -1.9941% -0.6139%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

window order by, built-in functions
                        time:   [7.4822 ms 7.5313 ms 7.5843 ms]
                        change: [-0.8959% +0.1569% +1.1590%] (p = 0.77 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

window partition by, u64_wide, aggregate functions
                        time:   [20.548 ms 20.664 ms 20.789 ms]
                        change: [+7.8065% +9.0739% +10.344%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

window partition by, u64_narrow, aggregate functions
                        time:   [7.2287 ms 7.2704 ms 7.3122 ms]
                        change: [+11.989% +13.130% +14.218%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild

window partition by, u64_wide, built-in functions
                        time:   [19.238 ms 19.360 ms 19.488 ms]
                        change: [+16.724% +18.053% +19.384%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

window partition by, u64_narrow, built-in functions
                        time:   [6.4687 ms 6.5140 ms 6.5591 ms]
                        change: [+8.1468% +9.2522% +10.283%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

window partition and order by, u64_wide, aggregate functions
                        time:   [27.575 ms 27.785 ms 28.013 ms]
                        change: [+9.4015% +10.538% +11.674%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

Benchmarking window partition and order by, u64_narrow, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.5s, or reduce sample count to 90.
window partition and order by, u64_narrow, aggregate functions
                        time:   [54.981 ms 55.324 ms 55.702 ms]
                        change: [-0.6417% +0.7593% +2.0563%] (p = 0.29 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

window partition and order by, u64_wide, built-in functions
                        time:   [25.673 ms 25.817 ms 25.969 ms]
                        change: [+7.6207% +8.7502% +9.8676%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

window partition and order by, u64_narrow, built-in functions
                        time:   [16.124 ms 16.211 ms 16.308 ms]
                        change: [+0.5055% +1.5834% +2.6646%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
```

File: datafusion/benches/window_query_sql.rs (hunk 1 above)
Comment (Jimexist) [reply datafusion/benches/window_query_sql.rs]: @Dandandan new results updated in the PR description.

repartition is indeed magical

**Comment by Dandandan** [comment]
Nice, almost 5 times improvement for some queries!

But do those have to do with the changes wrt sort or with regards to partition by?

File: datafusion/src/physical_plan/windows.rs [hunk 2]
```diff
@@ -412,11 +412,14 @@ impl ExecutionPlan for WindowAggExec {
...
-        Partitioning::UnknownPartitioning(1)
+        // because we can have repartitioning using the partition keys
+        // this would be either 1 or more than 1 depending on the presense of
+        // repartitioning
+        self.input.output_partitioning()
     }
 
     fn required_child_distribution(&self) -> Distribution {
-        Distribution::SinglePartition
+        Distribution::UnspecifiedDistribution
```
Comment (Dandandan) [inline datafusion/src/physical_plan/windows.rs:422]: Would this be correct with a window without any `partition by` clause?

In that case I think the required partitions should be 1, as the aggregate function can not be computed only over a part of the data.

**Comment by Jimexist** [comment]
> Nice, almost 5 times improvement for some queries!
> 
> But do those have to do with the changes wrt sort or with regards to partition by?

I don't expect that much of an improvement. let me redo the benchmark after #573 is merged.

File: datafusion/src/physical_plan/windows.rs (hunk 2 above)
Comment (Jimexist) [reply datafusion/src/physical_plan/windows.rs]: thanks! this is fixed.

**Comment by Jimexist** [comment]
latest against the master

```
window empty over, aggregate functions
                        time:   [38.093 ms 38.813 ms 39.557 ms]
                        change: [+7.9976% +10.144% +12.365%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

window empty over, built-in functions
                        time:   [28.492 ms 29.010 ms 29.571 ms]
                        change: [-9.6876% -6.5061% -3.1331%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

Benchmarking window order by, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 18.2s, or reduce sample count to 20.
window order by, aggregate functions
                        time:   [180.14 ms 182.30 ms 184.63 ms]
                        change: [-3.2247% -1.5082% +0.1867%] (p = 0.10 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  8 (8.00%) high mild
  2 (2.00%) high severe

Benchmarking window order by, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 16.7s, or reduce sample count to 20.
window order by, built-in functions
                        time:   [164.02 ms 165.72 ms 167.55 ms]
                        change: [-3.0666% -1.6099% -0.2141%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

Benchmarking window partition by, u64_wide, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 26.9s, or reduce sample count to 10.
window partition by, u64_wide, aggregate functions
                        time:   [279.31 ms 283.97 ms 288.78 ms]
                        change: [+5.6067% +7.9223% +10.288%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking window partition by, u64_narrow, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 13.2s, or reduce sample count to 30.
window partition by, u64_narrow, aggregate functions
                        time:   [119.82 ms 122.24 ms 124.91 ms]
                        change: [-10.403% -8.3916% -6.3109%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

Benchmarking window partition by, u64_wide, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 23.7s, or reduce sample count to 20.
window partition by, u64_wide, built-in functions
                        time:   [229.23 ms 232.16 ms 235.14 ms]
                        change: [+1.8582% +3.3754% +4.9304%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking window partition by, u64_narrow, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.9s, or reduce sample count to 40.
window partition by, u64_narrow, built-in functions
                        time:   [109.41 ms 111.44 ms 113.55 ms]
                        change: [-9.5580% -7.6470% -5.5888%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

Benchmarking window partition and order by, u64_wide, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 46.0s, or reduce sample count to 10.
window partition and order by, u64_wide, aggregate functions
                        time:   [451.08 ms 454.83 ms 458.61 ms]
                        change: [-8.3071% -6.4767% -4.6734%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking window partition and order by, u64_narrow, aggregate functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 60.7s, or reduce sample count to 10.
window partition and order by, u64_narrow, aggregate functions
                        time:   [564.17 ms 568.36 ms 572.64 ms]
                        change: [-17.391% -16.295% -15.200%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking window partition and order by, u64_wide, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 41.1s, or reduce sample count to 10.
window partition and order by, u64_wide, built-in functions
                        time:   [396.51 ms 401.62 ms 407.54 ms]
                        change: [-10.962% -8.9661% -6.8683%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

Benchmarking window partition and order by, u64_narrow, built-in functions: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 34.4s, or reduce sample count to 10.
window partition and order by, u64_narrow, built-in functions
                        time:   [331.09 ms 334.20 ms 337.40 ms]
                        change: [-16.892% -15.883% -14.890%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

@Dandandan turns out it's not that clear cut

**Review by Dandandan — APPROVED** [review]
👍

**Pull request merged**